File: lab-3-dynamodb/dynamoDBJavaLab/src/InfectionStatistics.java
// Copyright 2017 Amazon Web Services, Inc. or its affiliates. All rights reserved.

import com.amazonaws.AmazonClientException;
import com.amazonaws.AmazonServiceException;
import com.amazonaws.regions.Region;
import com.amazonaws.services.dynamodbv2.AmazonDynamoDBClient;
import com.amazonaws.services.dynamodbv2.document.DynamoDB;
import com.amazonaws.services.dynamodbv2.document.Item;
import com.amazonaws.services.dynamodbv2.document.ItemCollection;
import com.amazonaws.services.dynamodbv2.document.QueryOutcome;
import java.util.Iterator;

// The InfectionStatistics class queries the infections table and reports the total number of infections in a city.
public class InfectionStatistics {

  // City for which number of infections must be queried
  public static final String inputCity = "Reno";

  public static int itemCount = 0;

  private static final String INFECTIONS_TABLE_NAME = InfectionsTableCreator.INFECTIONS_TABLE_NAME;
  private static final String CITY_DATE_INDEX_NAME = InfectionsTableCreator.CITY_DATE_INDEX_NAME;
  private static final Region REGION = Utils.getRegion();

  private static DynamoDB dynamoDB = null;
  private static AmazonDynamoDBClient dynamoDBClient = null;

  public static void main(String[] args) throws Exception {

    try {
      // Instantiate DynamoDB client and object
      dynamoDBClient = new AmazonDynamoDBClient();
      dynamoDBClient.setRegion(REGION);
      dynamoDB = new DynamoDB(dynamoDBClient);

      // Query items for a city
      ItemCollection<QueryOutcome> items =
          queryCityRelatedItems(dynamoDB, INFECTIONS_TABLE_NAME, CITY_DATE_INDEX_NAME, inputCity);

      System.out.println("-------------------------------------------------------------");

      // Retrieve a reference to an iterator from the ItemCollection object returned by the query.
      Iterator<Item> itemIterator = items.iterator();
      Item item = null;
      // Iterate over items
      while (itemIterator.hasNext()) {
        // Count the number of items
        itemCount++;
        item = itemIterator.next();
        System.out.printf(
            "%s - %s - %s %n",
            item.getString("PatientId"), item.getString("City"), item.getString("Date"));
      }
      System.out.println("-------------------------------------------------------------");
      System.out.printf("Summary: Number of infections in %s city is %d %n", inputCity, itemCount);

    } catch (AmazonServiceException ase) {
      System.out.println("Error Message:    " + ase.getMessage());
      System.out.println("HTTP Status Code: " + ase.getStatusCode());
      System.out.println("AWS Error Code:   " + ase.getErrorCode());
      System.out.println("Error Type:       " + ase.getErrorType());
      System.out.println("Request ID:       " + ase.getRequestId());
    } catch (AmazonClientException ace) {
      System.out.println("Error Message: " + ace.getMessage());
    }
  }

  /**
   * Query the infections table to retrieve the items that relate to the given city
   *
   * @param dynamoDB              Instance of DynamoDB class
   * @param infectionsTableName   Table name
   * @param cityDateGsiName       Global secondary index
   * @param inputCity             City
   * @return                      Query results
   */
  public static ItemCollection<QueryOutcome> queryCityRelatedItems(
      DynamoDB dynamoDB, String infectionsTableName, String cityDateGsiName, String inputCity) {
    // STUDENT TODO 3: Replace the solution with your own code
    return Solution.queryCityRelatedItems(
        dynamoDB, infectionsTableName, cityDateGsiName, inputCity);
  }
}

File: lab-3-dynamodb/dynamoDBPythonLab/infection_statistics.py
# Copyright 2017 Amazon Web Services, Inc. or its affiliates. All rights
# reserved.

import boto3
from boto3.dynamodb.conditions import Key
import sys
from datetime import datetime, timedelta
import utils
import solution as dynamodb_solution

INFECTIONS_TABLE_NAME = utils.LAB_S3_INFECTIONS_TABLE_NAME
CITY_DATE_INDEX_NAME = "InfectionsByCityDate"


def query_by_city(city):
    print("City name is {0}".format(city))
    # Query Infections table based on the input city and count the number of
    # infections
    count_for_city = 0
    try:
        dynamodb = boto3.resource('dynamodb')
        recs = query_city_related_items(
            dynamodb, INFECTIONS_TABLE_NAME, CITY_DATE_INDEX_NAME, city)

        # Retrieves and prints from recs dictionary returned by the query.
        for rec in recs['Items']:
            print("\t", rec['PatientId'], rec['Date'])
        count_for_city = len(recs['Items'])
        print("Count of Infections in the city: {0}".format(count_for_city))
    except Exception as err:
        print("Error Message: {0}".format(err))
    return count_for_city


def query_city_related_items(dynamodb, infections_table_name, gsi_name, city):
    """Return the items returned by the query

    Keyword arguments:
    dynamodb -- DynamoDB resource
    infections_table_name -- Table name
    gsi_name -- Table index name
    city -- City name
    """

    # STUDENT TODO 3: Replace the solution with your own code
    return dynamodb_solution.query_city_related_items(
        dynamodb, infections_table_name, gsi_name, city)

if __name__ == '__main__':
    print("Querying items by city")
    query_by_city(city="Reno")
